File: convolve.py
#! /usr/bin/env python3

import sys
import time
import itertools

def tuplewise(n, iterable):
    """
    Take successive windows of an iterable, n at a time,
    moving the window by 1 each time.

    >>> list(tuplewise(2, [1, 2, 3, 4]))
    [(1, 2), (2, 3), (3, 4)]
    >>> list(tuplewise(3, [1, 2, 3, 4]))
    [(1, 2, 3), (2, 3, 4)]
    """
    iters = itertools.tee(iterable, n)
    for repeat, it in enumerate(iters):
        for i in range(repeat):
            next(it)
    return zip(*iters)


def iter_row(y, matrix):
    return iter(matrix[y])

def iter_column(x, matrix):
    for row in matrix:
        yield row[x]

def convolve_minmax(matrix):
    """
    Takes in a 2-D matrix, as a nested list. Computes the convolution of
    the matrix with the kernel [-1, 0, 1] vertically and horizontally. Returns
    4 values, (dx_max, dx_min, dy_max, dy_min) for the most extreme differences
    found by the convolution.

    >>> convolve_minmax([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    (2, 2, 6, 6)
    >>> convolve_minmax([[1, -2, 3], [-4, 5, -6], [7, -8, 9]])
    (2, -2, 6, -6)
    """
    # We don't actually store the dx and dy matrices, because we only need the
    # min and max. In fact, we only use a constant amount more memory than the
    # size of the original matrix.

    height = len(matrix)
    width = len(matrix[0])

    # X direction.
    dx_max = sys.float_info.min
    dx_min = sys.float_info.max
    for y in range(height):
        row = iter_row(y, matrix)
        for a, b, c in tuplewise(3, row):
            n = -a + c
            if n > dx_max:
                dx_max = n
            if n < dx_min:
                dx_min = n

    # Y direction.
    dy_max = sys.float_info.min
    dy_min = sys.float_info.max
    for x in range(width):
        column = iter_column(x, matrix)
        for a, b, c in tuplewise(3, column):
            n = -a + c
            if n > dy_max:
                dy_max = n
            if n < dy_min:
                dy_min = n

    return (dx_max, dx_min, dy_max, dy_min)

def timer(func):
    start = time.time()
    result = func()
    finish = time.time()
    return result, (finish - start)

if __name__ == '__main__':
    import random
    from textwrap import dedent

    # Take width and height from command line.
    width, height = sys.argv[1:3]
    width = int(width)
    height = int(height)

    # Generate a matrix of random values.
    matrix = [
        [random.random() for _ in range(width)]
        for _ in range(height)
    ]

    # Convolve and print maximums and minimums.
    result, elapsed = timer(lambda: convolve_minmax(matrix))

    dx_max, dx_min, dy_max, dy_min = result
    output = '''
        dx max {:.6f}
        dx min {:.6f}
        dy max {:.6f}
        dy min {:.6f}
        elapsed time: {:.4f}s
    '''.format(dx_max, dx_min, dy_max, dy_min, elapsed)
    output = dedent(output).strip()
    print(output)

File: convolve.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <sys/time.h>
#include <limits.h>

#ifdef __MACH__
#include <mach/clock.h>
#include <mach/mach.h>
#endif

long get_time_microseconds() {
    /* Adapted from
     * http://stackoverflow.com/questions/275004/c-timer-function-to-provide-time-in-nano-seconds
     */
    struct timespec ts;

#ifdef __MACH__ // OS X does not have clock_gettime, use clock_get_time
    clock_serv_t cclock;
    mach_timespec_t mts;
    host_get_clock_service(mach_host_self(), CALENDAR_CLOCK, &cclock);
    clock_get_time(cclock, &mts);
    mach_port_deallocate(mach_task_self(), cclock);
    ts.tv_sec = mts.tv_sec;
    ts.tv_nsec = mts.tv_nsec;
#else
    clock_gettime(CLOCK_REALTIME, &ts);
#endif

    return (long)ts.tv_sec * 1000 + ts.tv_nsec / 1000;
}


int main(int argc, char **argv) {
    /* Get command line arguments. */
    if (argc < 3) { return 1; }
    char *width_string = argv[1];
    char *height_string = argv[2];

    int width = atoi(width_string);
    int height = atoi(height_string);


    /* Generate a random matrix of size width x height. */
    srand(time(NULL));
    int x, y;
    int **matrix;
    matrix = malloc(height * sizeof(int*));
    for (y=0; y<height; y++) {
        matrix[y] = malloc(width * sizeof(int));
        for (x=0; x<width; x++) {
            matrix[y][x] = rand();
        }
    }

    /*
     * Find max and min convolution values for the kernel [-1, 0, 1], in both
     * the x and y directions.
     */
    long start, finish;
    start = get_time_microseconds();

    int n;
    int dx_max = INT_MIN;
    int dx_min = INT_MAX;
    int dy_max = INT_MIN;
    int dy_min = INT_MAX;
    for (y=0; y<height; y++) {
        for (x=0; x<width-2; x++) {
            n = -matrix[y][x] + matrix[y][x+2];
            if (n > dx_max) {
                dx_max = n;
            }
            if (n < dx_min) {
                dx_min = n;
            }
        }
    }
    for (x=0; x<width; x++) {
        for (y=0; y<height-2; y++) {
            n = -matrix[y][x] + matrix[y+2][x];
            if (n > dy_max) {
                dy_max = n;
            }
            if (n < dy_min) {
                dy_min = n;
            }
        }
    }

    finish = get_time_microseconds();

    /* Display results. */
    printf("dx_max %d\ndx_min %d\ndy_max %d\ndy_min %d\n", dx_max, dx_min, dy_max, dy_min);
    printf("elapsed %fs\n", (finish - start)/1000000.0);

    /* Free matrix memory */
    for (y=0; y<height; y++) {
        free(matrix[y]);
    }
    free(matrix);

    return 0;
}
